examples/DataApi-Water-And-Energy-Usage/template/modules/shared_data/src/main/scala/com/my/currency/shared_data/Main.scala

```diff
@@ -12,69 +12,218 @@ import org.http4s.HttpRoutes
 import org.tessellation.currency.dataApplication.dataApplication.DataApplicationValidationErrorOr
 import org.tessellation.currency.dataApplication.{DataApplicationValidationError, DataState, DataUpdate}
 import org.tessellation.security.signature.Signed
+import cats.syntax.all._
+import com.my.currency.shared_data.Data.{AggregatedUsage, EnergyUsage, WaterUsage}
+import com.my.currency.shared_data.Validations.{validateEnergyTimestampToNoDoubling, validateEnergyUsageUpdate, validateProvidedAddress, validateWaterTimestampToNoDoubling, validateWaterUsageUpdate}
+import monocle.syntax.all._
+import org.tessellation.schema.address.Address
+import org.tessellation.security.SecurityProvider
 
 import java.nio.charset.StandardCharsets
 
+object Errors {
+  case object EnergyNotPositive extends DataApplicationValidationError {
+    val message = "energy usage must be positive"
+  }
+
+  case object EnergyAlreadyUpdated extends DataApplicationValidationError {
+    val message = "Energy already updated, skipping"
+  }
+
+  case object WaterNotPositive extends DataApplicationValidationError {
+    val message = "water usage must be positive"
+  }
+
+  case object WaterAlreadyUpdated extends DataApplicationValidationError {
+    val message = "Energy already updated, skipping"
+  }
+
+  case object EmptyUpdate extends DataApplicationValidationError {
+    val message = "Provided an empty update"
+  }
+
+  case object InvalidAddress extends DataApplicationValidationError {
+    val message = "Provided address different than proof"
+  }
+}
+
+object Validations {
+  def validateEnergyUsageUpdate(energy: EnergyUsage): DataApplicationValidationErrorOr[Unit] =
+    if (energy.usage > 0L && energy.timestamp > 0L)
+      ().validNec
+    else
+      EnergyNotPositive.asInstanceOf[DataApplicationValidationError].invalidNec
+
+  def validateWaterUsageUpdate(water: WaterUsage): DataApplicationValidationErrorOr[Unit] =
+    if (water.usage > 0L && water.timestamp > 0L)
+      ().validNec
+    else
+      WaterNotPositive.asInstanceOf[DataApplicationValidationError].invalidNec
+
+  def validateEnergyTimestampToNoDoubling(maybeState: Option[AggregatedUsage], energy: EnergyUsage): DataApplicationValidationErrorOr[Unit] = {
```

Comment:
I think the intention here with this method name is to indicate that the validation prevents block doubling/double spends but I don't think anyone would understand that without more context. I would suggest changing this name to something more readable like `validateEnergyTimestamp` and leaving a comment that explains the implication of requiring a greater timestamp than the previous update.